pageserver/src/tenant/timeline.rs

```diff
@@ -708,19 +708,27 @@ impl Timeline {
         flags: EnumSet<CompactFlags>,
         ctx: &RequestContext,
     ) -> Result<(), CompactionError> {
-        let _g = self.compaction_lock.lock().await;
+        // most likely the cancellation token is from background task, but in tests it could be the
+        // request task as well.
+
+        let prepare = async move {
+            let guard = self.compaction_lock.lock().await;
+
+            let permit = super::tasks::concurrent_background_tasks_rate_limit_permit(
+                BackgroundLoopKind::Compaction,
+                ctx,
+            )
+            .await;
+
+            (guard, permit)
+        };
 
         // this wait probably never needs any "long time spent" logging, because we already nag if
         // compaction task goes over it's period (20s) which is quite often in production.
-        let _permit = match super::tasks::concurrent_background_tasks_rate_limit(
-            BackgroundLoopKind::Compaction,
-            ctx,
-            cancel,
-        )
-        .await
-        {
-            Ok(permit) => permit,
-            Err(RateLimitError::Cancelled) => return Ok(()),
+        let (_guard, _permit) = tokio::select! {
+            tuple = prepare => { tuple },
+            _ = self.cancel.cancelled() => return Ok(()),
+            _ = cancel.cancelled() => return Ok(()),
```

Comment:
In the common case, compaction_lock is uncontended while background rate limit might be.
So, should acquire them in opposite order: background limit first, then compaction_lock.

In the "want to delete timeline but waiting for compaction" case, your introduction of cancellation sensitivity here would make us bail out immediatly, regardless of how long either compaction_lock or rate limit take to acquire. So, no risk in reversing the order, i.e., no risk in doing background rate limit before compaction lock.

Can be a separate PR though, feel free to turn into follow-up issue.

Reply:
I was considering this as well. When I added this in #5108 or in a call I am pretty sure we discussed it. I was thinking acquiring local resources first which should be uncontented should be first. But yeah, unsure if that is a strong argument anymore.

Reply:
Thinking about this more, not taking the local resource first would also open up us to taking multiple slots from the global resource.